lib/lvol/lvol.c

```diff
@@ -1378,6 +1433,124 @@ spdk_lvol_create_snapshot(struct spdk_lvol *origlvol, const char *snapshot_name,
 				lvol_create_cb, req);
 }
 
+struct xattrs_ctx {
+	struct spdk_lvol *newlvol;
+	const char *const *xattrs_external;
+};
+
+static void
+lvol_get_xattr_value_ext(void *xattr_ctx, const char *name,
+			 const void **value, size_t *value_len)
+{
+	struct xattrs_ctx *ctx = xattr_ctx;
+	size_t i;
+
+	lvol_get_xattr_value(ctx->newlvol, name, value, value_len);
+
+	if (*value == NULL || *value_len == 0) {
+		/* Search for external xattr */
+		for (i = 0; ctx->xattrs_external[i]; i++) {
+			/* Skip xattrs value, we search xattr names, which are in even positions */
+			if (i % 2 != 0) {
+				continue;
+			}
+
+			if (strcmp(name, ctx->xattrs_external[i]) == 0) {
+				*value = ctx->xattrs_external[i + 1];
+				*value_len = strlen(ctx->xattrs_external[i + 1]) + 1;
+				return;
+			}
+		}
+
+		*value = NULL;
+		*value_len = 0;
+	}
+}
+
+void
+spdk_lvol_create_snapshot_with_xattrs(struct spdk_lvol *origlvol, const char *snapshot_name,
+				      const char *const *xattrs_external, size_t xattrs_external_num,
+				      spdk_lvol_op_with_handle_complete cb_fn, void *cb_arg)
+{
+	struct spdk_lvol_store *lvs;
+	struct spdk_lvol *newlvol;
+	struct spdk_blob *origblob;
+	struct spdk_lvol_with_handle_req *req;
+	struct spdk_blob_xattr_opts snapshot_xattrs;
+	struct xattrs_ctx xattr_ctx;
+	char **xattr_names;
+	char *xattr_names_internal[] = {LVOL_NAME, "uuid", LVOL_CREATION_TIME};
+	int rc;
+
+	if (origlvol == NULL) {
+		SPDK_INFOLOG(lvol, "Lvol not provided.\n");
+		cb_fn(cb_arg, NULL, -EINVAL);
+		return;
+	}
+
+	if (xattrs_external == NULL) {
+		SPDK_WARNLOG("Xattr not provided.\n");
+		return spdk_lvol_create_snapshot(origlvol, snapshot_name, cb_fn, cb_arg);
+	}
+
+	if (xattrs_external_num % 2 != 0) {
+		SPDK_ERRLOG("Xattr list must contain couples of key-value\n");
+		cb_fn(cb_arg, NULL, -EINVAL);
+		return;
+	}
+
+	xattr_names = lvol_create_xattrs_names((const char *const *)&xattr_names_internal, xattrs_external,
+					       SPDK_COUNTOF(xattr_names_internal), xattrs_external_num);
+	if (xattr_names == NULL) {
+		cb_fn(cb_arg, NULL, -ENOMEM);
```

Comment:
don't we need to return here?

Reply:
Yes, you are right David, thanks. Out of memory error are not covered by unit test ...

Reply:
Your comment forced me to think better to possible errors, I have inserted a duplicate function to  make a copy of external xattrs that come from RPC. It can be dangerous to pass them directly into the blob layer.

Reply:
@innobead Could you resolve this requested change? Otherwise the PR can't be merged

Reply:
You can resolve that if that has been fixed.

Reply:
This comment is resolved, but there is still 1 change requested by you that block the merge